File: spectrum_overlay.py
"""spectrum_overlay.py — a built-in, transparent bottom-up audio spectrum
that overlays the mission-graph area (the left + center columns).

This replaces relying on an external `glava` window: it's drawn directly
inside the deck as the top-most layer over that region, click-through, and
kept low-alpha enough that the mission graph and the pipes texture stay
clearly visible underneath the bars.

Enabled with CYBERDECK_GLAVA=1 (see cyberdeck.py).
"""
import math, struct, subprocess, time
from PySide6.QtWidgets import QWidget
from PySide6.QtCore import Qt, QTimer, Signal, QThread
from PySide6.QtGui import QPainter, QColor, QLinearGradient

# Cyan, matching the previous `glava --desktop -m graph` look — a single
# hue rather than the hue-cycling rainbow, so it reads as one coherent
# glow rising up from the baseline instead of a multicolor equalizer.
CYAN_LOW  = QColor(30, 60, 100)     # One Dark blue dim
CYAN_HIGH = QColor(97, 175, 239)    # One Dark blue


def clamp(v, lo=0.0, hi=1.0):
    return max(lo, min(hi, v))


class AudioThread(QThread):
    """Reads live audio levels via PulseAudio's `parec`. Falls back to a
    gentle idle wave if parec/pulseaudio isn't available, so the overlay
    never just crashes or sits frozen."""
    levels = Signal(list)

    def __init__(self, n=48):
        super().__init__()
        self.n = n
        self.data = [0.0] * n
        self._run = True

    def run(self):
        try:
            # Record what's PLAYING (the default sink's monitor), not the
            # microphone — parec with no -d captures the default source.
            monitor = None
            try:
                sink = subprocess.check_output(
                    ["pactl", "get-default-sink"], timeout=2, text=True).strip()
                if sink:
                    monitor = sink + ".monitor"
            except Exception:
                pass
            cmd = ["parec", "--format=float32le", "--rate=44100", "--channels=1"]
            if monitor:
                cmd += ["-d", monitor]
            proc = subprocess.Popen(
                cmd, stdout=subprocess.PIPE, stderr=subprocess.DEVNULL,
            )
            while self._run:
                raw = proc.stdout.read(self.n * 4)
                if not raw:
                    break
                vals = struct.unpack(f"{self.n}f", raw)
                for i, v in enumerate(vals):
                    self.data[i] = self.data[i] * 0.6 + min(abs(v) * 4.0, 1.0) * 0.4
                self.levels.emit(self.data[:])
        except FileNotFoundError:
            while self._run:
                t = time.time()
                for i in range(self.n):
                    v = math.sin(t * 1.4 + i * 0.25) * 0.15 + 0.15
                    self.data[i] = self.data[i] * 0.7 + clamp(v) * 0.3
                self.levels.emit(self.data[:])
                time.sleep(0.05)

    def stop(self):
        self._run = False
        self.wait(1000)


class SpectrumOverlay(QWidget):
    """Transparent, click-through equalizer bars anchored to the bottom of
    the region it's given, rising upward with the music."""

    def __init__(self, parent=None, n=48):
        super().__init__(parent)
        self.setAttribute(Qt.WA_TranslucentBackground)
        self.setAttribute(Qt.WA_TransparentForMouseEvents)
        self.n = n
        self.raw = [0.0] * n
        self.smooth = [0.0] * n

        self._audio = AudioThread(n)
        self._audio.levels.connect(self._on_audio)
        self._audio.start()

        self._t = QTimer(self)
        self._t.timeout.connect(self._tick)
        self._t.start(33)

    def _on_audio(self, data):
        self.raw = data

    def _tick(self):
        for i in range(self.n):
            r = self.raw[i] if i < len(self.raw) else 0.0
            self.smooth[i] = self.smooth[i] * 0.78 + r * 0.22
        self.update()

    def paintEvent(self, e):
        p = QPainter(self)
        p.setRenderHint(QPainter.Antialiasing)
        w, h = self.width(), self.height()
        bw = max(4, w // self.n - 2)
        tw = self.n * (bw + 2)
        sx = (w - tw) // 2
        for i in range(self.n):
            bh = int(self.smooth[i] * h * 0.55)
            if bh <= 0:
                continue
            x = sx + i * (bw + 2)
            y = h - bh
            # Low alpha throughout — this rides on top of the graph and
            # pipes texture, so it needs to stay see-through rather than
            # painting a solid wall of bars over them. Single cyan hue,
            # dim at the base and brighter toward the tip.
            g = QLinearGradient(x, y + bh, x, y)
            g.setColorAt(0.0, QColor(CYAN_LOW.red(), CYAN_LOW.green(), CYAN_LOW.blue(), 35))
            g.setColorAt(1.0, QColor(CYAN_HIGH.red(), CYAN_HIGH.green(), CYAN_HIGH.blue(), 130))
            p.setPen(Qt.NoPen)
            p.setBrush(g)
            p.drawRoundedRect(x, y, bw, bh, 2, 2)
        p.end()

    def closeEvent(self, e):
        self._audio.stop()
        self._audio.deleteLater()

File: cpp-sworddeck/src/spectrum_overlay.cpp
#include "spectrum_overlay.h"

#include <QPainter>
#include <QLinearGradient>
#include <QProcess>
#include <QFile>
#include <QMutexLocker>
#include <cmath>
#include <ctime>

static const QColor CYAN_LOW(30, 60, 100);
static const QColor CYAN_HIGH(97, 175, 239);

AudioThread::AudioThread(int n, QObject *parent)
    : QThread(parent), m_n(n), m_data(n, 0.0) {}

void AudioThread::stop() { m_running = false; wait(1000); }

void AudioThread::run() {
    // Try parec
    QString monitor;
    QProcess pactl;
    pactl.start("pactl", {"get-default-sink"});
    if (pactl.waitForFinished(2000)) {
        QString sink = QString::fromLocal8Bit(pactl.readAllStandardOutput()).trimmed();
        if (!sink.isEmpty()) monitor = sink + ".monitor";
    }

    QStringList args = {"--format=float32le", "--rate=44100", "--channels=1"};
    if (!monitor.isEmpty()) args << "-d" << monitor;

    QProcess proc;
    proc.start("parec", args);
    if (proc.waitForStarted(1000)) {
        while (m_running) {
            QByteArray raw = proc.read(m_n * 4);
            if (raw.size() < m_n * 4) break;
            const float *vals = reinterpret_cast<const float*>(raw.constData());
            for (int i = 0; i < m_n; i++) {
                m_data[i] = m_data[i] * 0.6 + qMin(std::abs(vals[i]) * 4.0, 1.0) * 0.4;
            }
            emit levels(m_data);
        }
    } else {
        // Fallback: idle wave
        while (m_running) {
            double t = time(nullptr);
            for (int i = 0; i < m_n; i++) {
                double v = std::sin(t * 1.4 + i * 0.25) * 0.15 + 0.15;
                m_data[i] = m_data[i] * 0.7 + qBound(0.0, v, 1.0) * 0.3;
            }
            emit levels(m_data);
            msleep(50);
        }
    }
}

SpectrumOverlay::SpectrumOverlay(QWidget *parent, int n)
    : QWidget(parent), m_n(n), m_raw(n, 0.0), m_smooth(n, 0.0)
{
    setAttribute(Qt::WA_TranslucentBackground);
    setAttribute(Qt::WA_TransparentForMouseEvents);

    m_audio = new AudioThread(n, this);
    connect(m_audio, &AudioThread::levels, this, &SpectrumOverlay::onAudio);
    m_audio->start();

    connect(&m_timer, &QTimer::timeout, this, &SpectrumOverlay::tick);
    m_timer.start(33);
}

void SpectrumOverlay::onAudio(const QVector<double> &data) { m_raw = data; }

void SpectrumOverlay::tick() {
    for (int i = 0; i < m_n; i++)
        m_smooth[i] = m_smooth[i] * 0.78 + (i < m_raw.size() ? m_raw[i] : 0.0) * 0.22;
    update();
}

void SpectrumOverlay::paintEvent(QPaintEvent *) {
    QPainter p(this);
    p.setRenderHint(QPainter::Antialiasing);
    int w = width(), h = height();
    int bw = qMax(4, w / m_n - 2);
    int tw = m_n * (bw + 2);
    int sx = (w - tw) / 2;

    for (int i = 0; i < m_n; i++) {
        int bh = (int)(m_smooth[i] * h * 0.55);
        if (bh <= 0) continue;
        int x = sx + i * (bw + 2);
        int y = h - bh;
        QLinearGradient g(x, y + bh, x, y);
        g.setColorAt(0.0, QColor(CYAN_LOW.red(), CYAN_LOW.green(), CYAN_LOW.blue(), 35));
        g.setColorAt(1.0, QColor(CYAN_HIGH.red(), CYAN_HIGH.green(), CYAN_HIGH.blue(), 130));
        p.setPen(Qt::NoPen);
        p.setBrush(g);
        p.drawRoundedRect(x, y, bw, bh, 2, 2);
    }
    p.end();
}

void SpectrumOverlay::closeEvent(QCloseEvent *e) {
    m_audio->stop();
    m_audio->deleteLater();
    QWidget::closeEvent(e);
}
